### PR TITLE
[WIP] Adding b-tensor encoding scripts like compute_fodf/frf and metrics

### PR DESCRIPTION
**Pull request by karanphil**
Here are three scripts `scil_compute_memsmt_frf.py`, `scil_compute_memsmt_fofd.py` and `scil_compute_divide.py` enabling the use of tensor-valued dMRI data. The scripts are almost Scilpy ready, but the side modules are messy. This needs to be cleaned up before the PR is ready.

### CODE REVIEW

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/8/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/10/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/11/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/12/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/13/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/14/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/15/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/16/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/17/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by karanphil** [comment]
@arnaudbore if you want to test the three scripts, here is some data.

https://usherbrooke-my.sharepoint.com/:f:/g/personal/karp2601_usherbrooke_ca/EqLMY8wcU85EjGT4J19kAhABBgsdVS3EF3QgVsdDlG23jA?e=TTlnYQ

Here are examples of command lines for the scripts.

`scil_compute_memsmt_frf.py wm_frf.txt gm_frf.txt csf_frf.txt --in_dwi_linear dwi_lin_eddy_corrected_n4.nii.gz --in_bval_linear dwi_lin.bval --in_bvec_linear dwi_lin.bvec --in_dwi_spherical dwi_sph_eddy_corrected_n4.nii.gz --in_bval_spherical dwi_sph.bval --in_bvec_spherical dwi_sph.bvec --mask mask.nii.gz --mask_wm wm_mask.nii.gz --mask_gm gm_mask.nii.gz --mask_csf csf_mask.nii.gz --roi_center 45 51 29 --wm_frf_mask wm_frf_mask.nii.gz --gm_frf_mask gm_frf_mask.nii.gz --csf_frf_mask csf_frf_mask.nii.gz -f`

`scil_compute_memsmt_fodf.py wm_frf.txt gm_frf.txt csf_frf.txt --in_dwi_linear dwi_lin_eddy_corrected_n4.nii.gz --in_bval_linear dwi_lin.bval --in_bvec_linear dwi_lin.bvec --in_dwi_spherical dwi_sph_eddy_corrected_n4.nii.gz --in_bval_spherical dwi_sph.bval --in_bvec_spherical dwi_sph.bvec --mask mask.nii.gz --processes 8 -f`

`scil_compute_divide.py --in_dwi_linear dwi_lin_eddy_corrected_n4.nii.gz --in_bval_linear dwi_lin.bval --in_bvec_linear dwi_lin.bvec --in_dwi_spherical dwi_sph_eddy_corrected_n4.nii.gz --in_bval_spherical dwi_sph.bval --in_bvec_spherical dwi_sph.bvec --mask mask.nii.gz --fa fa.nii.gz  -f`

**Review by mdesco — COMMENTED** [review]
I have launched scil_compute_divide.py with only linear data it ran... I don't think it should @karanphil , no?

Otherwise, @arnaudbore we approve for now and move my small comments to issues that @karanphil  can correct when he gets back

**Comment by mdesco** [comment]
@karanphil what is the status on this PR? Welcome back! :)

**Comment by karanphil** [comment]
> @karanphil what is the status on this PR? Welcome back! :)

Thanks! I'm looking into it. I'll come back to you when it's ready!

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/20/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/21/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/29/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/30/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/32/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/33/

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/38/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/39/

**Comment by karanphil** [comment]
@arnaudbore @mdesco I cleaned up the code and changed a few things. I think it's ready for testing. Here is some data.

https://usherbrooke-my.sharepoint.com/:f:/g/personal/karp2601_usherbrooke_ca/EqLMY8wcU85EjGT4J19kAhABBgsdVS3EF3QgVsdDlG23jA?e=TTlnYQ

Here are examples of command lines for the scripts.

`scil_compute_memsmt_frf.py wm_frf.txt gm_frf.txt csf_frf.txt --in_dwis dwi_lin_eddy_corrected_n4.nii.gz dwi_sph_eddy_corrected_n4.nii.gz --in_bvals dwi_lin.bval dwi_sph.bval --in_bvecs dwi_lin.bvec dwi_sph.bvec --in_bdeltas 1 0 --mask mask.nii.gz --mask_wm wm_mask.nii.gz --mask_gm gm_mask.nii.gz --mask_csf csf_mask.nii.gz --roi_center 45 51 29 --wm_frf_mask wm_frf_mask.nii.gz --gm_frf_mask gm_frf_mask.nii.gz --csf_frf_mask csf_frf_mask.nii.gz -f`

`scil_compute_memsmt_fodf.py wm_frf.txt gm_frf.txt csf_frf.txt --in_dwis dwi_lin_eddy_corrected_n4.nii.gz dwi_sph_eddy_corrected_n4.nii.gz --in_bvals dwi_lin.bval dwi_sph.bval --in_bvecs dwi_lin.bvec dwi_sph.bvec --in_bdeltas 1 0 --mask mask.nii.gz --processes 8 -f`

`scil_compute_divide.py --in_dwis dwi_lin_eddy_corrected_n4.nii.gz dwi_sph_eddy_corrected_n4.nii.gz --in_bvals dwi_lin.bval dwi_sph.bval --in_bvecs dwi_lin.bvec dwi_sph.bvec --in_bdeltas 1 0 --mask mask.nii.gz --fa fa.nii.gz -f`

Thanks!

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/47/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/48/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/49/

**Comment by scil-admin** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/50/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/51/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/52/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/53/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/54/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/55/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/56/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/57/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/58/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/59/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/60/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/61/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/62/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/63/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/64/

**Comment by arnaudbore** [comment]
Build Failed :boom: 
https://jenkins.scil.usherbrooke.ca/job/scilpy/job/PR-425/65/

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by karanphil** [comment]
> @arnaudbore @mdesco I cleaned up the code and changed a few things. I think it's ready for testing. Here is some data.
> 
> https://usherbrooke-my.sharepoint.com/:f:/g/personal/karp2601_usherbrooke_ca/EqLMY8wcU85EjGT4J19kAhABBgsdVS3EF3QgVsdDlG23jA?e=TTlnYQ
> 
> Here are examples of command lines for the scripts.
> 
> `scil_compute_memsmt_frf.py wm_frf.txt gm_frf.txt csf_frf.txt --in_dwis dwi_lin_eddy_corrected_n4.nii.gz dwi_sph_eddy_corrected_n4.nii.gz --in_bvals dwi_lin.bval dwi_sph.bval --in_bvecs dwi_lin_eddy_corrected.eddy_rotated_bvecs dwi_sph_eddy_corrected.eddy_rotated_bvecs --in_bdeltas 1 0 --mask mean_b0_lin_topup_bet_mask.nii.gz --mask_wm wm_mask.nii.gz --mask_gm gm_mask.nii.gz --mask_csf csf_mask.nii.gz --roi_center 45 51 29 --wm_frf_mask wm_frf_mask.nii.gz --gm_frf_mask gm_frf_mask.nii.gz --csf_frf_mask csf_frf_mask.nii.gz -f`
> 
> `scil_compute_memsmt_fodf.py wm_frf.txt gm_frf.txt csf_frf.txt --in_dwis dwi_lin_eddy_corrected_n4.nii.gz dwi_sph_eddy_corrected_n4.nii.gz --in_bvals dwi_lin.bval dwi_sph.bval --in_bvecs dwi_lin_eddy_corrected.eddy_rotated_bvecs dwi_sph_eddy_corrected.eddy_rotated_bvecs --in_bdeltas 1 0 --mask mean_b0_lin_topup_bet_mask.nii.gz --processes 8 -f`
> 
> `scil_compute_divide.py --in_dwis dwi_lin_eddy_corrected_n4.nii.gz dwi_sph_eddy_corrected_n4.nii.gz --in_bvals dwi_lin.bval dwi_sph.bval --in_bvecs dwi_lin_eddy_corrected.eddy_rotated_bvecs dwi_sph_eddy_corrected.eddy_rotated_bvecs --in_bdeltas 1 0 --mask mean_b0_lin_topup_bet_mask.nii.gz --fa fa.nii.gz --processes 8 -f`
> 
> Thanks!

Here is the data.

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

File: scripts/scil_compute_memsmt_fodf.py [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+
+    if args.vf_rgb:
+        vf_rgb = vf / np.max(vf) * 255
+        vf_rgb = np.clip(vf_rgb, 0, 255)
+        nib.save(nib.Nifti1Image(vf_rgb.astype(np.uint8),
+                                 affine), args.vf_rgb)
+
+
+if __name__ == "__main__":
+    main()
```
[the remark this answers is not in the document]
Comment (mdesco) [reply scripts/scil_compute_memsmt_fodf.py:375]: /Users/desm2239/my_env/scilpy_p10/lib/python3.10/site-packages/cvxpy/problems/problem.py:1385: UserWarning: Solution may be inaccurate. Try another solver, adjusting the solver settings, or solve with verbose=True for more information.
  warnings.warn(

File: scripts/scil_compute_divide.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    if args.mk_i:
+        nib.save(nib.Nifti1Image(MK_I.astype(np.float32), affine), args.mk_i)
+    if args.mk_a:
+        nib.save(nib.Nifti1Image(MK_A.astype(np.float32), affine), args.mk_a)
+    if args.mk_t:
+        nib.save(nib.Nifti1Image(MK_T.astype(np.float32), affine), args.mk_t)
+
+
+if __name__ == "__main__":
+    main()
```
Comment (mdesco) [inline scripts/scil_compute_divide.py:226]: (CVXPY) Mar 16 03:22:46 PM: Encountered unexpected exception importing solver SCS:
ImportError('dlopen(/Users/desm2239/my_env/scilpy_p10/lib/python3.10/site-packages/_scs_direct.cpython-310-darwin.so, 0x0002): Symbol not found: ___emutls_get_address\n  Referenced from: <17C4D491-3B6D-344A-8328-189690282A3A> /Users/desm2239/my_env/scilpy_p10/lib/python3.10/site-packages/scs/.dylibs/libopenblasp-r0.3.21.dylib\n  Expected in:     <E0C76B2B-0EC6-364E-AACD-42C7AAE73F23> /Applications/freesurfer/lib/gcc/lib/libgomp.1.dylib')

**Review by mdesco — COMMENTED** [review]
Minor suggestions for @karanphil 
All the CVX errors seem like warnings, the output is created and it looks fine.

File: docs/source/documentation/btensor_scripts.rst [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+Then, you should compute the fODFs and volume fractions. The following command will save a fODF file for each tissue and a volume fractions file. This script should run in about 1-2 hours for a full brain.
+::
+
+    scil_compute_memsmt_fodf.py wm_frf.txt gm_frf.txt csf_frf.txt --in_dwis dwi_linear.nii.gz dwi_planar.nii.gz dwi_spherical.nii.gz --in_bvals dwi_linear.bval dwi_planar.bval dwi_spherical.bval --in_bvecs dwi_linear.bvec dwi_planar.bvec dwi_spherical.bvec --in_bdeltas 1 -0.5 0 --mask mask.nii.gz --processes 8 -f
+
+If you want to do DIVIDE with b-tensor data, you should use the following command. It will save files for the MD, uFA, OP, MK_I, MK_A and MK_T. This script should run in about 1-2 hours for a full brain.
+::
+
+    scil_compute_divide.py --in_dwis dwi_linear.nii.gz dwi_planar.nii.gz dwi_spherical.nii.gz --in_bvals dwi_linear.bval dwi_planar.bval dwi_spherical.bval --in_bvecs dwi_linear.bvec dwi_planar.bvec dwi_spherical.bvec --in_bdeltas 1 -0.5 0 --mask mask.nii.gz --fa fa.nii.gz --processes 8 -f
```
[the remark this answers is not in the document]
Comment (karanphil) [reply docs/source/documentation/btensor_scripts.rst:18]: I don't have the planar data, this example was very general.

File: scripts/scil_compute_memsmt_frf.py [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+        response_wm = responses[0]
+        iso_responses = np.concatenate((response_csf[:, :3],
+                                        response_gm[:, :3]), axis=1)
+        responses = np.concatenate((iso_responses, response_wm[:, :3]), axis=1)
+        frf_table = np.vstack((bvals, responses.T)).T
+        np.savetxt(args.frf_table, frf_table)
+
+
+if __name__ == "__main__":
+    main()
```
[the remark this answers is not in the document]
Comment (karanphil) [reply scripts/scil_compute_memsmt_frf.py:304]: I never got this error with Linux. This is weird. @arnaudbore what do you think? It looks like cvxpy tries to use some package that Max doesn't have.

**Comment by arnaudbore** [comment]
Build passed ! Good Job :beers: !

**Review by mdesco — APPROVED** [review]
Tested on Laurent's mac. The errors and warnings are not there. 
Output are the same as mine. So, correct the small edits and good to merge. 

Good job!

**Review by arnaudbore — APPROVED** [review]

**Pull request merged**